File: mobile/lib/features/water/domain/water_entry.dart
/// A logged water intake event (`/water-entries`).
class WaterEntry {
  const WaterEntry({
    required this.id,
    required this.consumedAt,
    required this.volumeLiters,
    this.sourceId,
    this.sourceName,
  });

  final int id;
  final DateTime consumedAt;
  final double volumeLiters;
  final int? sourceId;
  final String? sourceName;

  factory WaterEntry.fromJson(Map<String, dynamic> json) {
    return WaterEntry(
      id: json['id'] as int,
      consumedAt: DateTime.parse(json['consumedAt'] as String),
      volumeLiters: (json['volumeLiters'] as num).toDouble(),
      sourceId: json['sourceId'] as int?,
      sourceName: json['sourceName'] as String?,
    );
  }
}

File: backend/src/main/java/com/lifey/water/WaterEntry.java
package com.lifey.water;

import com.lifey.common.domain.SyncableEntity;
import com.lifey.user.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.Instant;

/**
 * A logged water intake event. {@link #volumeLiters} is a snapshot taken at
 * creation time (copied from the {@link WaterSource} if one was used) so
 * editing or deleting a source later never changes past entries — the FK is
 * {@code on delete set null} for the same reason, purely informational.
 */
@Getter
@Setter
@Entity
@Table(name = "water_entries")
public class WaterEntry extends SyncableEntity {

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "water_source_id")
    private WaterSource waterSource;

    @Column(name = "volume_liters", nullable = false)
    private double volumeLiters;

    @Column(name = "consumed_at", nullable = false)
    private Instant consumedAt;
}
